### PR TITLE
Land/Water Raise/Lower Game Actions

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Another 4 since they are so similar.

### CODE REVIEW

**Comment by ZehMatt** [comment]
Sadly I found some differences when it comes to errors. It seems that water is fine altho raising/lowering land has a few differences. It only happens if you use the grid something like 3x3 and collide with something such as a ride.
develop:
openrct2 com_2019-03-09_13-29-21e1403d4f598b7656c
the pr:
openrct2_2019-03-09_13-31-10064939078ee745b5d
Functionality wise, it seems to be all in order.

**Comment by duncanspumpkin** [comment]
Your issue is fixed in the rebase. Was a mistake I made with `LandSetHeightAction`

**Comment by ZehMatt** [comment]
Great, I'll do some tests and go over the code.

**Comment by ZehMatt** [comment]
Another issue that I found is that the error message is shown on host and client in some cases, but this is unrelated to the PR. Ideally errors are shown for the callee and not everyone.

File: src/openrct2/actions/WaterLowerAction.hpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        res->ExpenditureType = RCT_EXPENDITURE_TYPE_LANDSCAPING;
+
+        uint8_t minHeight = GetLowestHeight();
+        bool hasChanged = false;
+        for (int32_t y = validRange.GetTop(); y <= validRange.GetBottom(); y += 32)
+        {
+            for (int32_t x = validRange.GetLeft(); x <= validRange.GetRight(); x += 32)
+            {
+                TileElement* tileElement = map_get_surface_element_at(x / 32, y / 32);
+                if (tileElement != nullptr)
```
Comment (ZehMatt) [inline src/openrct2/actions/WaterLowerAction.hpp]: Maybe continue here too.

File: src/openrct2/actions/WaterLowerAction.hpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+private:
+    uint8_t GetLowestHeight() const
+    {
+        uint8_t minHeight{ 0 };
+        for (int32_t y = _range.GetTop(); y <= _range.GetBottom(); y += 32)
+        {
+            for (int32_t x = _range.GetLeft(); x <= _range.GetRight(); x += 32)
+            {
+                TileElement* tile_element = map_get_surface_element_at({ x, y });
+                if (tile_element != nullptr)
```
Comment (ZehMatt) [inline src/openrct2/actions/WaterLowerAction.hpp]: Could use continue here too.

File: src/openrct2/actions/WaterRaiseAction.hpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        res->ExpenditureType = RCT_EXPENDITURE_TYPE_LANDSCAPING;
+
+        uint8_t maxHeight = GetHighestHeight();
+        bool hasChanged = false;
+        for (int32_t y = validRange.GetTop(); y <= validRange.GetBottom(); y += 32)
+        {
+            for (int32_t x = validRange.GetLeft(); x <= validRange.GetRight(); x += 32)
+            {
+                TileElement* tileElement = map_get_surface_element_at(x / 32, y / 32);
+                if (tileElement != nullptr)
```
Comment (ZehMatt) [inline src/openrct2/actions/WaterRaiseAction.hpp]: Maybe continue here too.

File: src/openrct2/actions/WaterRaiseAction.hpp [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+private:
+    uint8_t GetHighestHeight() const
+    {
+        uint8_t maxHeight{ 255 };
+        for (int32_t y = _range.GetTop(); y <= _range.GetBottom(); y += 32)
+        {
+            for (int32_t x = _range.GetLeft(); x <= _range.GetRight(); x += 32)
+            {
+                TileElement* tile_element = map_get_surface_element_at({ x, y });
+                if (tile_element != nullptr)
```
Comment (ZehMatt) [inline src/openrct2/actions/WaterRaiseAction.hpp]: Maybe continue here too.

File: src/openrct2/actions/LandLowerAction.hpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+        uint8_t maxHeight = map_get_highest_land_height(
+            validRange.GetLeft(), validRange.GetRight(), validRange.GetTop(), validRange.GetBottom());
+
+        for (int32_t y = validRange.GetTop(); y <= validRange.GetBottom(); y += 32)
+        {
+            for (int32_t x = validRange.GetLeft(); x <= validRange.GetRight(); x += 32)
+            {
+                TileElement* tileElement = map_get_surface_element_at(x / 32, y / 32);
+                if (tileElement != nullptr)
```
Comment (ZehMatt) [inline src/openrct2/actions/LandLowerAction.hpp]: Maybe its better to do 
```
if(tileElement == nullptr)
    continue;
```
Not a fan of scope pyramids.

File: src/openrct2/actions/LandRaiseAction.hpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+        uint8_t minHeight = map_get_lowest_land_height(
+            validRange.GetLeft(), validRange.GetRight(), validRange.GetTop(), validRange.GetBottom());
+
+        for (int32_t y = validRange.GetTop(); y <= validRange.GetBottom(); y += 32)
+        {
+            for (int32_t x = validRange.GetLeft(); x <= validRange.GetRight(); x += 32)
+            {
+                TileElement* tileElement = map_get_surface_element_at(x / 32, y / 32);
+                if (tileElement != nullptr)
```
Comment (ZehMatt) [inline src/openrct2/actions/LandRaiseAction.hpp]: Maybe continue here too.

**Comment by duncanspumpkin** [comment]
Part of #8878

**Pull request closed without merging**